generators/csharp/sdk/src/oauth/OauthTokenProviderGenerator.ts

```diff
@@ -0,0 +1,267 @@
+import { csharp, CSharpFile, FileGenerator } from "@fern-api/csharp-codegen";
+import { join, RelativeFilePath } from "@fern-api/fs-utils";
+import {
+    EndpointReference,
+    ExampleRequestBody,
+    http,
+    HttpEndpoint,
+    HttpService,
+    Name,
+    OAuthScheme,
+    ObjectProperty,
+    ResponseProperty
+} from "@fern-fern/ir-sdk/api";
+import { SdkCustomConfigSchema } from "../SdkCustomConfig";
+import { SdkGeneratorContext } from "../SdkGeneratorContext";
+
+export declare namespace OauthTokenProviderGenerator {
+    interface Args {
+        scheme: OAuthScheme;
+        context: SdkGeneratorContext;
+    }
+}
+
+const OAUTH_TOKEN_PROVIDER_CLASS_NAME = "OAuthTokenProvider";
+
+export class OauthTokenProviderGenerator extends FileGenerator<CSharpFile, SdkCustomConfigSchema, SdkGeneratorContext> {
+    private classReference: csharp.ClassReference;
+    private scheme: OAuthScheme;
+    private tokenEndpointHttpService: HttpService;
+    private tokenEndpointReference: EndpointReference;
+    private tokenEndpoint: HttpEndpoint;
+
+    private static readonly BUFFER_IN_MINUTES_FIELD = csharp.field({
+        access: "private",
+        const_: true,
+        name: "BufferInMinutes",
+        type: csharp.Type.double(),
+        initializer: csharp.codeblock("2")
+    });
+
+    private static readonly ACCESS_TOKEN_FIELD = csharp.field({
+        access: "private",
+        name: "_accessToken",
+        type: csharp.Type.optional(csharp.Type.string())
+    });
+
+    private static readonly EXPIRES_AT_FIELD = csharp.field({
+        access: "private",
+        name: "_expiresAt",
+        type: csharp.Type.optional(csharp.Type.dateTime())
+    });
+
+    private static readonly CLIENT_ID_FIELD = csharp.field({
+        access: "private",
+        name: "_clientId",
+        type: csharp.Type.string()
+    });
+
+    private static readonly CLIENT_SECRET_FIELD = csharp.field({
+        access: "private",
+        name: "_clientSecret",
+        type: csharp.Type.string()
+    });
+
+    constructor({ context, scheme }: OauthTokenProviderGenerator.Args) {
+        super(context);
+        this.scheme = scheme;
+        this.classReference = csharp.classReference({
+            namespace: this.context.getCoreNamespace(),
+            name: OAUTH_TOKEN_PROVIDER_CLASS_NAME
+        });
+
+        this.tokenEndpointReference = this.scheme.configuration.tokenEndpoint.endpointReference;
+        this.tokenEndpointHttpService = this.context.getHttpServiceOrThrow(this.tokenEndpointReference.serviceId);
+        const httpEndpoint = this.tokenEndpointHttpService.endpoints.find(
+            (endpoint) => endpoint.id === this.tokenEndpointReference.endpointId
+        );
+
+        if (httpEndpoint == null) {
+            throw new Error(`Failed to find token endpoint ${this.tokenEndpointReference.endpointId}`);
+        }
+        this.tokenEndpoint = httpEndpoint;
```

Comment:
nit: for symmetry with the `getHttpServiceOrThrow` method call above, a private helper function that read a little better.

```suggestion
        this.tokenEndpointReference = this.scheme.configuration.tokenEndpoint.endpointReference;
        this.tokenEndpointHttpService = this.context.getHttpServiceOrThrow(this.tokenEndpointReference.serviceId);
        this.tokenEndpoint = this.resolveTokenEndpointOrThrow(this.tokenEndpointReference);
```